src/arweave/composite-client.ts

```diff
@@ -726,67 +828,106 @@ export class ArweaveCompositeClient
-            timedOut: false,
-          };
-        } catch (error: any) {
-          let canceled = false;
-          let timedOut = false;
-
-          if (axios.isAxiosError(error)) {
-            timedOut = error.code === 'ECONNABORTED';
-            canceled = error.code === 'ERR_CANCELED';
-          }
-
-          this.log.warn('Failed to broadcast chunk:', {
-            message: error.message,
-            stack: error.stack,
-          });
-
+        } else {
           failureCount++;
 
           metrics.arweaveChunkPostCounter.inc({
             endpoint: url,
             status: 'fail',
+            role: 'primary',
           });
-
-          return {
-            success: false,
-            statusCode: error.response?.status,
-            canceled,
-            timedOut,
-          };
         }
+
+        return response;
       }),
     );
 
+    let results = primaryResults;
+    let secondarySuccessCount = 0;
+
+    if (this.secondaryChunkPostUrls.length > 0) {
+      const secondaryResults: Promise<BroadcastChunkResponses>[] = [];
+
+      const shuffledSecondaryChunkPostUrls = this.secondaryChunkPostUrls.sort(
+        () => Math.random() - 0.5,
+      );
+
+      const limit = pLimit(this.secondaryChunkPostConcurrency);
+      for (const url of shuffledSecondaryChunkPostUrls) {
+        if (secondarySuccessCount >= this.secondaryChunkPostMinSuccessCount)
+          break;
+
+        const circuitBreaker = this.secondaryChunkPostCircuitBreakers[url];
+        const task = limit(async () => {
+          try {
+            const response = await circuitBreaker.fire({
+              url,
+              chunk,
+              abortTimeout,
+              responseTimeout,
+              originAndHopsHeaders,
+            });
+
+            if (response.success) {
+              secondarySuccessCount++;
+              metrics.arweaveChunkPostCounter.inc({
+                endpoint: url,
+                status: 'success',
+                role: 'secondary',
+              });
+            } else {
+              failureCount++;
+              metrics.arweaveChunkPostCounter.inc({
+                endpoint: url,
+                status: 'fail',
+                role: 'secondary',
+              });
+            }
+
+            return response;
+          } catch (error: any) {
+            failureCount++;
+            this.log.debug('Secondary chunk broadcast failed:', { error });
+            metrics.arweaveChunkPostCounter.inc({
+              endpoint: url,
+              status: 'fail',
+              role: 'secondary',
+            });
+            return {
+              success: false,
+              statusCode: error?.response?.status,
+              canceled: error.code === 'ERR_CANCELED',
+              timedOut: error.code === 'ECONNABORTED',
+            };
+          }
+        });
+
+        secondaryResults.push(task);
+      }
+
+      results = [...results, ...(await Promise.all(secondaryResults))];
```

Comment:
As mentioned below, let's adjust this to avoid waiting for results before returning a response.